enterprise/internal/campaigns/resolvers/changeset_connection_test.go

```diff
@@ -72,63 +79,157 @@ func TestChangesetConnectionResolver(t *testing.T) {
 		externalBranch:      "merged-pr",
 		externalState:       campaigns.ChangesetExternalStateMerged,
 		publicationState:    campaigns.ChangesetPublicationStatePublished,
+		externalReviewState: campaigns.ChangesetReviewStatePending,
+		ownedByCampaign:     campaign.ID,
+		campaign:            campaign.ID,
+	})
+	changeset4 := createChangeset(t, ctx, store, testChangesetOpts{
+		repo:                inaccessibleRepo.ID,
+		externalServiceType: "github",
+		externalID:          "987651",
+		externalBranch:      "open-hidden-pr",
+		externalState:       campaigns.ChangesetExternalStateOpen,
+		publicationState:    campaigns.ChangesetPublicationStatePublished,
+		externalReviewState: campaigns.ChangesetReviewStatePending,
 		ownedByCampaign:     campaign.ID,
 		campaign:            campaign.ID,
 	})
 
 	addChangeset(t, ctx, store, campaign, changeset1.ID)
 	addChangeset(t, ctx, store, campaign, changeset2.ID)
 	addChangeset(t, ctx, store, campaign, changeset3.ID)
+	addChangeset(t, ctx, store, campaign, changeset4.ID)
 
 	s, err := graphqlbackend.NewSchema(&Resolver{store: store}, nil, nil)
 	if err != nil {
 		t.Fatal(err)
 	}
 
 	campaignApiID := string(campaigns.MarshalCampaignID(campaign.ID))
+	nodes := []apitest.Changeset{
+		{
+			Typename:   "ExternalChangeset",
+			ID:         string(marshalChangesetID(changeset1.ID)),
+			Repository: apitest.Repository{Name: repo.Name},
+		},
+		{
+			Typename:   "ExternalChangeset",
+			ID:         string(marshalChangesetID(changeset2.ID)),
+			Repository: apitest.Repository{Name: repo.Name},
+		},
+		{
+			Typename:   "ExternalChangeset",
+			ID:         string(marshalChangesetID(changeset3.ID)),
+			Repository: apitest.Repository{Name: repo.Name},
+		},
+		{
+			Typename: "HiddenExternalChangeset",
+			ID:       string(marshalChangesetID(changeset4.ID)),
+		},
+	}
 
-	input := map[string]interface{}{"campaign": campaignApiID}
-	var response struct{ Node apitest.Campaign }
-	apitest.MustExec(ctx, t, s, input, &response, queryChangesetConnection)
-
-	wantChangesets := apitest.ChangesetConnection{
-		Stats: apitest.ChangesetConnectionStats{
-			Unpublished: 1,
-			Open:        1,
-			Merged:      1,
-			Closed:      0,
-			Total:       3,
+	tests := []struct {
+		First           int
+		unsafeOpts      bool
+		wantHasNextPage bool
+		wantTotalCount  int
+		wantOpen        int
+		wantNodes       []apitest.Changeset
+	}{
+		{
+			First:           1,
+			wantHasNextPage: true,
+			wantTotalCount:  4,
+			wantOpen:        2,
+			wantNodes:       nodes[:1],
+		},
+		{
+			First:           2,
+			wantHasNextPage: true,
+			wantTotalCount:  4,
+			wantOpen:        2,
+			wantNodes:       nodes[:2],
+		},
+		{
+			First:           3,
+			wantHasNextPage: true,
+			wantTotalCount:  4,
+			wantOpen:        2,
+			wantNodes:       nodes[:3],
+		},
```

Comment:
How wide would these be on a single line? 🤔 
```suggestion
	{ First: 3, wantHasNextPage: true, wantTotalCount: 4, wantOpen: 2, wantNodes: nodes[:3] },
```